src/urllib3/contrib/emscripten/fetch.py

```diff
@@ -0,0 +1,477 @@
+            if js.Atomics.wait(self.int_buffer, 0, 0, self.timeout) == "timed-out":
+                raise _TimeoutError
+            data_len = self.int_buffer[0]
+            if data_len > 0:
+                self.read_len = data_len
+                self.read_pos = 0
+            elif data_len == ERROR_EXCEPTION:
+                raise _StreamingError
+            else:
+                # EOF, free the buffers and return zero
+                self.read_len = 0
+                self.read_pos = 0
+                self.int_buffer = None
+                self.byte_buffer = None
+                return 0
+        # copy from int32array to python bytes
+        ret_length = min(self.read_len, len(memoryview(byte_obj)))
+        subarray = self.byte_buffer.subarray(
+            self.read_pos, self.read_pos + ret_length
+        ).to_py()
+        memoryview(byte_obj)[0:ret_length] = subarray
+        self.read_len -= ret_length
+        self.read_pos += ret_length
+        return ret_length
+
+
+class _StreamingFetcher:
+    def __init__(self) -> None:
+        # make web-worker and data buffer on startup
+        self.streaming_ready = False
+
+        dataBlob = js.Blob.new(
+            [_STREAMING_WORKER_CODE], _obj_from_dict({"type": "application/javascript"})
+        )
+
+        def promise_resolver(res: JsProxy, rej: JsProxy) -> None:
+            def onMsg(e: JsProxy) -> None:
+                self.streaming_ready = True
+                res(e)
+
+            def onErr(e: JsProxy) -> None:
+                rej(e)
+
+            self._worker.onmessage = onMsg
+            self._worker.onerror = onErr
+
+        dataURL = js.URL.createObjectURL(dataBlob)
+        self._worker = js.globalThis.Worker.new(dataURL)
+        self._worker_ready_promise = js.globalThis.Promise.new(promise_resolver)
+
+    def send(self, request: EmscriptenRequest) -> EmscriptenResponse:
+        headers = {
+            k: v for k, v in request.headers.items() if k not in HEADERS_TO_IGNORE
+        }
+
+        body = request.body
+        fetch_data = {"headers": headers, "body": to_js(body), "method": request.method}
+        # start the request off in the worker
+        timeout = int(1000 * request.timeout) if request.timeout > 0 else None
+        shared_buffer = js.SharedArrayBuffer.new(1048576)
+        int_buffer = js.Int32Array.new(shared_buffer)
+        byte_buffer = js.Uint8Array.new(shared_buffer, 8)
+
+        js.Atomics.store(int_buffer, 0, 0)
+        js.Atomics.notify(int_buffer, 0)
+        absolute_url = js.URL.new(request.url, js.location).href
+        self._worker.postMessage(
+            _obj_from_dict(
+                {
+                    "buffer": shared_buffer,
+                    "url": absolute_url,
+                    "fetchParams": fetch_data,
+                }
+            )
+        )
+        # wait for the worker to send something
+        js.Atomics.wait(int_buffer, 0, 0, timeout)
+        if int_buffer[0] == 0:
```

Comment:
Is this zero a magic value too? Should it be stored in a similar way to the others?